platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,709 @@
+the new roster. The Hashgraph module will decide in what round the new roster becomes active. Let us define `N` as the
+number of rounds from the latest consensus round at which the candidate roster will become active. This number will be
+configured with the same value for all nodes in the network.
+
+For example, suppose Execution has just finished handling round 10, and has buffered up round 11. The most recent round
+that the Hashgraph module produced was round 11. Suppose `N` is 10. If the call to `nextRound` from Execution at the
+end of round 10 were to supply a new Roster (`R`). The Hashgraph module will assign `R` to become active as of round 22
+(`latest_consensus_round + N + 1`). It will then work on producing consensus round 12. Once produced, metadata
+indicating that `R` will be used for rounds starting at 22 is passed to Gossip, Event Intake, Event Creator, and
+Execution. The roster used for round 22 may also be included in this metadata.
+
+There is a critical relationship between the value of N, the latency for adopting a new roster, and the number of future
+events held in memory on the node. Remember that each event has a birth-round. Let us define another variable called
+`Z`, such that `Z <= N`. Any event with `birth_round > latest_consensus_round + z` is considered a "far future" event,
+and will be dropped by the birth-round filtering logic in Event Intake. Any event with
+`birth_round > latest_consensus_round` is just a "future" event.
+
+The smaller the value of `Z`, the fewer events are held in memory on the node. However, events with birth-rounds just
+larger than `Z` will be dropped and have to be retrieved on demand with sync gossip, rather than through the broadcast
+tree. Sync is slower and more laborious than broadcast gossip, so the efficiency of the node will go down with low
+values of `Z`. Larger values of `Z` means more events in memory, but it also means more "smoothing" in gossip and
+in handling any performance spikes.
+
+On the other hand, the larger the value of `N`, the larger the latency between the round we know about a new roster, and
+the round at which it can be actually used. Ideally, the value of `N` would be small, like 3, but we may find that 3 is
+too small a number for `Z`.
+
+Each node may select its own value of `Z`, so long as it is less than or equal to `N`. But all nodes must use the same
+value for `N`, or they will ISS since they will assign different rosters to different rounds. The value of `N` is not
+defined here, but will need some investigation. Unfortunately, this number appears to require "just being chosen" and
+cannot be deterministically dynamically computed based on network conditions.
+
+Note: when the roster is applied to some future round `latest_consensus_round + N + 1`, if such a round already exists
+in the hashgraph, then that round and any future rounds must be removed and the events reapplied in order given the new
```

Comment:
I believe it's more accurate to say that the events have their metadata recalculated with the new roster. The term `remove` is misleading, IMO.

Reply:
This section might not be aware of how the meta data is recalculated in each consensus round.   I tried to explain it in the next comment down below.

Reply:
@lpetrovic05, please correct me if I am wrong in my understanding of the expected behavior of virtual voting in the consensus algorithm. 

Suppose the pending consensus round is X with active roster A.   Round X+N has a new roster B set. 

It is possible for [virtual voting](https://docs.hedera.com/hedera/core-concepts/hashgraph-consensus-algorithms/virtual-voting) to create "future rounds" that instantiate meta data with round created of X+N and higher, but all the weights and nodes and events considered come from Roster A and not Roster B, even for rounds X+N and higher. 

Roster B is only relevant once consensus starts calculating `pending consensus round` X+N.  

@rbair23, this needs a much deeper discussion if we do what you suggest and apply the new roster early to virtual voting in `pending consensus round` X.  This wasn't the form of dynamism that was discussed with Leemon.  

This behavior of virtual-voting also has implications in how much change we can tolerate in the roster. The deltas have to be small enough that virtual voting is still viable under the previous roster.

Reply:
@edward-swirldslabs if I understood correctly, what you said is correct. let me rephrase it to avoid any potential misunderstanding:

- We only calculate metadata using 1 roster at a time
- When we are calculating round `X`, we are using roster `R`
- Once we decide fame for round `X`, we re-calculate all event metadata using roster `R+1` in order to decide fame for `X+1`. (`R+1` could be the same as `R` or different)
- If `R+1` is not available to us at this point, there is nothing we can do, we have to wait for it
- So there is no removal of events, just a recalculation of metadata with the latest roster every round

Reply:
Based on the conversation yesterday, I think that part about removing future rounds should not be here. We don't even calculate fame for a round until we have a roster, so no removal is needed.

Reply:
I think changing the vocab of `future round` to `voting round` in this section will help.  

Do Z or N needs to be formal config items?  How are they determined if not?

Reply:
I've removed the section in question, as suggested by Lazar. I don't think a vocab change is sufficient here, it sounds like the concept of "removing" rounds simply isn't relevant.

> Do Z or N needs to be formal config items?

@edward-swirldslabs What do you mean by "formal config item"? I think how we will choose these values is left intentionally vague

Reply:
Right, keep forgetting that this proposal is supposed to be high level with more to follow with specifics.